### ISSUE TITLE
Bindlist in Postgres

### ISSUE THREAD

**Issue by anderssv** [jdbi/jdbi#1550]
We're issuing a select query with a `WHERE x IN (<list>)` clause with bindlist. 

For our case _we halve the query time_ when we change it to be `WHERE x IN (VALUES ('id1'), ('id2'), ('id3'))` . But we're doing basic string manipulation to achieve that, and would rather not. :)

Is there support in JDBI for this kind of query? Or can we extend it in some way to have a custom binding? I've tried reading through some of the code but didn't find any extension points I could use. 

Any pointers? How would you go about this?

The postgres feature: https://www.postgresql.org/docs/11/sql-values.html

Thanks!

**Comment by TheRealMarnes** [comment]
So in essence, rather than having your list values joined by `,`, you want to be able to specify a custom prefix, infix, and suffix for BindList to use, right?

**Comment by TheRealMarnes** [comment]
I think we can patch `SqlStatement#bindThis(String, List)` to support this:
![image](https://user-images.githubusercontent.com/3950300/60172252-99455c80-980c-11e9-92fa-a64dbcf1e706.png)

The extra strings can be passed to the method from the fluent API, and `@BindList` can get extra `prefix = ""`, `infix = ","`, `suffix = ""` parameters too. You could specify `prefix = "values (", infix = "), (", suffix = ")"` to achieve this.

At a glance, I'm up for supporting this. Opinions from @stevenschlansker @qualidafial ?

**Comment by qualidafial** [comment]
Have you tried binding the list as an array parameter (using just regular `@Bind` instead of `@BindList`) and use the `WHERE X = ANY(:param)`?

**Comment by aro** [comment]
@qualidafial  We just tested your suggestion. It works, but it is as slow as doing `WHERE x IN (<param>)`. 

Trying to bind using `WHERE x IN (VALUES :param)` or `WHERE x IN (VALUES <param>)` fails with an exception (_org.postgresql.util.PSQLException: ERROR: syntax error at or near "$2"_).

**Comment by TheRealMarnes** [comment]
I find it a bit strange that just adding the word `VALUES` before your value list makes any difference in performance. Does it just create different memory structures or activate different processing, so that it's a performance improvement when used under the right circumstances (large lists)?

**Comment by anderssv** [comment]
I'm not really proficient at the details but simple tests we've done on our DB shows a 2x improvement. 

This is the article we're leaning on that also describes performance boost when using VALUES: https://www.datadoghq.com/blog/100x-faster-postgres-performance-by-changing-1-line/

We also found similar advice on Stackoverflow: https://dba.stackexchange.com/questions/91247/optimizing-a-postgres-query-with-a-large-in/91254#91254


Regards,
Anders,

**Comment by TheRealMarnes** [comment]
We could perhaps make it a `jdbi3-postgres` specialization of `bindList` to render with `VALUES`, maybe with a count threshold if performance actually suffers with small lists (or so users of an old db version can disable it with `MAX_VALUE`). Feels more first-class feature, more user-friendly that way than making users do `@BindList(prefix = "VALUES(", infix = "), (", suffix = ")")`

**Comment by anderssv** [comment]
Yeah, it depends a bit on how you like to support it. Oh, BTW: We're using the fluent interface and not annotations so I have little input on the actual annotations. :)

Looking through the DataDog article now I see (right at the end) that after Postgres 9.3 their original problem was fixed so they would get better performance again when doing "ANY(ARRAY[...])" instead of VALUES. So that means there might be two things to support in the Postgres extension...

The prefix/suffix solution would avoid supporting multiple ways of doing it. 

But if you support it in the library as a specific thing I think it's best suited as a per query parameter? Maybe per binding, but I can see how that can complicate the API quite a bit. 

If it's per query you don't really need an threshold either, you just specify that you want to bind arrays with the "values method", "array method" or "plain" (default).

**Comment by jayd16** [comment]
I'm not sure about any of this performance talk but I would really like a way to access the `(VALUES <list>)` syntax in general.  You can use that syntax as a literal table and join against it.  This can be used differently than `IN (<list>)` as you can join against multiple rows with the same value.

**Comment by stevenschlansker** [comment]
What version of Postgres do you use when you observe this impressive speedup?  there was some important work on planning prepared statements around the Postgres 9 days, IIRC.

The gist of it is, when you prepare a statement like `= ANY(?)` there is always one parameter, and the statement is prepared and reused by the driver since it never changes.  So one query plan is prepared whether there is 1 element in your array or 10,000 elements.

The improvement was to make the executed query plan better match the actual shape of the actual argument provided at execute time.  So, if you are on an old Postgres, you may find the gap between the approaches disappears with a server + driver update.

But this could be a fine feature to add if people would use it.  Getting the escaping done right is possibly tricky...

**Comment by anderssv** [comment]
Hey. We were using Postgres 11 at the time. Running on RDS though, but other info from the articles above suggests it's not related to that. I'm not exactly sure about the exact version, but probably 11.2 or something like that. :) We're at 11.4 now.

**Comment by mmurfin87** [comment]
Bumping this because I'm running into issues trying to support this today. My use case is deduplicating some data to prevent reprocessing data that's already been seen and put in the database. I'm using postgres 14.

The query presented to JDBI is:
`SELECT link FROM (VALUES <linkList>) AS newlinks (link) EXCEPT SELECT link FROM oldlinks`

Assuming I provide linkList as something akin to 
`['one.com', 'two.com']` 
then I would expect the query to expand out to be 
`SELECT link FROM (VALUES ('one.com'), ('two.com')) AS feedlinks (link) EXCEPT SELECT link FROM feed_items`

This uses the same sql-values feature of postgres that was linked above. I'll link again for postgres 14. https://www.postgresql.org/docs/14/sql-values.html

**Comment by maxs-rose** [comment]
I know this is an older issue now but I have just come across it today did this ever get implemented?

**Comment by mmurfin87** [comment]
Bumping this again because I totally forgot about this issue but ran into it anew the other day.

It's still an annoying blocker, but I was able to basically accomplish what I want with either BindBeanList or BindMethodsList and wrapping the list argument in either a bean or a Supplier<>.

So for instance, while this doesn't work as intended:
```
@SqlQuery("SELECT * FROM (VALUES <valueList>) AS three_row_one_col")
public List<String> selectList(@BindList Iterable<String> valueList);
```

This actually works as intended:
```
@SqlQuery("SELECT * FROM (VALUES <valueList>) AS three_row_one_col")
public List<String> selectList(@BindMethodsList(methodName={"get"}) Iterable<Supplier<String>> valueList);
```

And Calling it just becomes a little weird:
```
final List<Supplier<String>> valueList = List.of("one", "two", "three").stream().map(str -> (Supplier<String>)() -> str).toList();
repository.selectList(valueList);
```
